### PR TITLE
BindToClose issue fixes

### PR DESCRIPTION
**Pull request by Grargror**
Addressed Issue #124 by moving the BindToClose function to a separate callback which is set to nil once the player leaves.

Addressed Issue #125 by using a BindableEvent to trigger the final save of player's data instead of setting the player object's parent to nil.  Promises seem to be an elegant way to handle the AncestryChanged and the new BindableEvent's connection.

Addressed Issue #126 by wrapping the BindToClose function in a pcall to prevent errors if the game shuts down before the BindToClose function has finished.

### CODE REVIEW

**Comment by Kampfkarren** [comment]
`Parent = nil` is from https://github.com/Kampfkarren/Roblox/issues/45, can you look over that and see if the discussion there is still accurate to now? I believe servers shutting down not firing PlayerRemoving, but firing AncestryChanged because of this.

**Comment by Grargror** [comment]
I took a look at #45. My understanding is that `Player.Parent = nil` was added in order to immediately fire the AncestryChanged event (which calls the anonymous function to save the player's data) when the game starts shutting down.

I believe this behavior is correctly reproduced in the commit which fixes #125.  For that commit, I added a BindableEvent (`bindToCloseEvent`) which is fired when the game starts to shut down (replacing `Player.Parent = nil`).  Firing this event resolves the `Promise.race` starting on Line 571 and will immediately save the data.

In short, here's my interpretation of the behavior:
* Old behavior: `Player.Parent = nil` fires`AncestryChanged` --> Saves player's data
* New behavior; ``bindToCloseEvent:Fire()` resolves `Promise.race` --> Saves player's data

**Comment by Kampfkarren** [comment]
Sounds good, can you double check with a server that shuts down that it still saves?

File: DataStore2/init.lua [hunk 1]
```diff
@@ -544,42 +543,58 @@ function DataStore2.__call(_, dataStoreName, player)
...
+		game:BindToClose(function()
+			if bindToCloseCallback == nil then
+				return
+			end
+	
+			bindToCloseCallback()
+		end)
 	end)
+	if success == false then
+		warn(errorMessage)
```
Comment (Kampfkarren) [inline DataStore2/init.lua:573]:
```suggestion
		warn("DataStore2 could not BindToClose", errorMessage)
```

Needs more info.

File: DataStore2/init.lua [hunk 2]
```diff
@@ -544,42 +543,58 @@ function DataStore2.__call(_, dataStoreName, player)
...
+	end
 
-	local playerLeavingConnection
-	playerLeavingConnection = player.AncestryChanged:Connect(function()
-		if player:IsDescendantOf(game) then return end
-		playerLeavingConnection:Disconnect()
+	Promise.race({
+		Promise.fromEvent(bindToCloseEvent.Event),
+		Promise.fromEvent(player.AncestryChanged, function()
+			return player:IsDescendantOf(game) == false
```
Comment (Kampfkarren) [inline DataStore2/init.lua:579]:
```suggestion
			return not player:IsDescendantOf(game)
```

File: DataStore2/init.lua [hunk 3]
```diff
@@ -544,42 +543,58 @@ function DataStore2.__call(_, dataStoreName, player)
...
+	local saveFinishedEvent, isSaveFinished = Instance.new("BindableEvent"), false
+	local bindToCloseEvent = Instance.new("BindableEvent")
 
-	game:BindToClose(function()
-		if not fired then
-			spawn(function()
-				player.Parent = nil -- Forces AncestryChanged to fire and save the data
-			end)
+	local bindToCloseCallback = function()
+		if isSaveFinished == false then
```
Comment (Kampfkarren) [inline DataStore2/init.lua:550]:
```suggestion
		if not isSaveFinished then
```

File: DataStore2/init.lua [hunk 4]
```diff
@@ -544,42 +543,58 @@ function DataStore2.__call(_, dataStoreName, player)
...
 		end)
 
-		delay(40, function() --Give a long delay for people who haven't figured out the cache :^(
+		--Give a long delay for people who haven't figured out the cache :^(
+		return Promise.delay(40):andThen(function() 
 			DataStoreCache[player] = nil
+			bindToCloseCallback = nil
 		end)
-	end)
+	end):catch(warn)
```
Comment (Kampfkarren) [inline DataStore2/init.lua:597]: Wrap this in an anonymous function that warns and gives the source, like the other `warn`.

**Comment by Grargror** [comment]
> 
> 
> Sounds good, can you double check with a server that shuts down that it still saves?

Yes, player data still appears to be saving properly in a server that is shut down.  Here's a video showing a shut down, and showing leaving the game.  Both cases save properly.

https://user-images.githubusercontent.com/22422247/118387239-9ad3c580-b5da-11eb-90dc-582eeba0b983.mp4

File: DataStore2/init.lua (hunk 4 above)
Comment (Grargror) [reply DataStore2/init.lua]: I added more info to the warning.  I wasn't quite sure what to tell the user though since this promise chain should never error.  `Promise.fromEvent` can only resolve as far as I know, so the `Promise.race` should never error.  Same with `Promise.delay`.

Comment (Kampfkarren) [reply DataStore2/init.lua]: Then you shouldn't need this at all, and we would rather this fatally error since something has gone terribly wrong, right?

Comment (Grargror) [reply DataStore2/init.lua]: Good point.  Removed in commit 91ade9e.

**Comment by Grargror** [comment]
Commit 5000de1 replaces the spawn function in the BindToClose callback function with `Promise.defer` (the "spiritual replacement for spawn").  Also added a comment to explain that the deferred code should continue to be deferred in order to avoid a race condition between connecting and firing the `saveFinishedEvent`.

**Comment by Kampfkarren** [comment]
Tests passed, thanks!

**Comment by Darkshepherd** [comment]
This is great work, thank you!
